### PR TITLE
Forms: Introduce new Switch component

### PR DESCRIPTION
**Pull request by dprokop**
Ref #19855 

Implements new Switch component:
image

### Note for Reviewer
The focus style is slightly different compared to the design, as we decided to resign from 1px gap between outline and the element for *all form elements*. This makes the focus ring blend with the background when the Switch is in `checked` state in Dark theme. 

Moreover, I'm not 100% sure if the outline should be visible at all when the actual Switch is focused - as it makes the outline show up when user clicks on the element. This behaviour feels wrong to me, but I'm looking forward to hear your opinion on that

### CODE REVIEW

File: packages/grafana-ui/src/components/Forms/Switch.tsx [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  disabled?: boolean;
+  onChange?: (e: React.SyntheticEvent<HTMLButtonElement>, checked: boolean) => void;
+}
+
+export const getSwitchStyles = stylesFactory((theme: GrafanaTheme) => {
+  return {
+    slider: cx(
+      css`
+        width: 32px;
+        height: 16px;
```
Comment (peterholmberg) [inline packages/grafana-ui/src/components/Forms/Switch.tsx]: No variables for these?

**Review by peterholmberg — COMMENTED** [review]
Looks good, just a question regarding variables.

**Comment by jessover9000** [comment]
Generally this is working fine. I do agree about both points regarding focus visibility on click as well as the missing gap/identical color between switch and focus effect.

Focus visibility: In the current form styles we're kind of taking the easy way out there by having identical hover and focus styles. I was wondering for a second whether focus should be for tab/keyboard only, but for smartphone users as well, the feedback is important and helpful to know for sure that you haven't missed the element. so for accessibility reasons I'm leaning towards keeping it.

Missing gap/identical color: I haven't questioned your recommendation to ditch the gap before, so this is a bit late, but just to make sure: I reckon the transparent outline solution doesn't work for us/has compatibility issues? (first button of last codepen) https://dev.to/elizabethschafer/designing-button-focus-states-for-better-usability-gm2 
Also, looking at the design, maybe there's been a mix-up between light and dark theme – the focus border should be = hover color, not enabled color :) which means that in dark theme, the focus border should be the darker blue and in light theme the lighter blue. Fixing that could help with the same color issue!

**Comment by dprokop** [comment]
> Missing gap/identical color: I haven't questioned your recommendation to ditch the gap before, so this is a bit late, but just to make sure: I reckon the transparent outline solution doesn't work for us/has compatibility issues? (first button of last codepen) https://dev.to/elizabethschafer/designing-button-focus-states-for-better-usability-gm2

Hm, actually the first solution in that example could work for us. cc @peterholmberg have you tried this approach?

**Comment by dprokop** [comment]
@jessover9000 I've created quick test of that approach, I think it works pretty well! You can test it out from this branch: https://github.com/grafana/grafana/tree/forms/focus-outline

**Comment by peterholmberg** [comment]
@dprokop 
> Hm, actually the first solution in that example could work for us. cc @peterholmberg have you tried this approach?

> The problem with this solution though is that it uses two box shadows, on to imitate the gap - this one needs to have some color set, can't be transparent. So we can assign i.e. page bg color, but I'm not 100% sure this is going to work in all case

Not with double box-shadows no, seems it might be worth a try.

**Review by torkelo — CHANGES_REQUESTED** [review]
Great start! 

1) There is a delay between after I click and it switches, so feels a bit laggy
2) I think the switch should animate the transition, both the movement of the dot and backgrounds color: 
Something like:

background-color 0.40s ease, transform 0.30s cubic-bezier(0.170, 0.895, 0.30, 1.265);

**Comment by dprokop** [comment]
> There is a delay between after I click and it switches, so feels a bit laggy

That's because the easing function used for transition - will fix that!

> I think the switch should animate the transition, both the movement of the dot and backgrounds color:
Something like:

👍

**Comment by dprokop** [comment]
@jessover9000 I have merged the updated focus style into this branch and applied changes suggested by you. Can you please try it again?

**Comment by jessover9000** [comment]
> @jessover9000 I have merged the updated focus style into this branch and applied changes suggested by you. Can you please try it again?

Oh yeeesss I love it! Well done :)

**Review by jessover9000 — APPROVED** [review]

**Review by peterholmberg — APPROVED** [review]
Looks good to me

**Review by torkelo — COMMENTED** [review]
Tested the latest version and the animation looks good but found a major problem. Same background checked / unchecked state, so impossible to know when it's unchecked or not :)

**Comment by dprokop** [comment]
Thanks @torkelo, good point. @jessover9000 any input on that, design-wise?

**Comment by torkelo** [comment]
I would start with this design: 
image 

And give that spin and see how it feels / look in action.

**Comment by torkelo** [comment]
The only thing that I do not like about that now that look at it is that the hover state makes it darker, usually hover states increase brightness (at least in a dark theme)

**Comment by jessover9000** [comment]
> The only thing that I do not like about that now that look at it is that the hover state makes it darker, usually hover states increase brightness (at least in a dark theme)

I think that's because Patrick wanted the regular state without interaction to be the highest possible contrast. In the light theme it's the other way round and the hover state makes it lighter. I don't think it's a problem to do it this way round, the main purpose of the hover effect is to emphasize the interactive nature of the event, nudge the user into clicking it. So I don't mind the way the colors behave now – especially since I like that the focus ring and the main switch color are not the same color and only become the same on hover.

**Comment by torkelo** [comment]
Yea, agree. It's not important, better that is looks the best in the non hover state :)

**Comment by dprokop** [comment]
@torkelo @jessover9000 that's totally my bad. I have considered the design on the right to be a `disabled` state, rather than OFF state. Sorry about that, I'll update this soon.

**Comment by dprokop** [comment]
@torkelo now the colors are updated

**Review by torkelo — APPROVED** [review]
Looks good now!

**Pull request merged**